hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterJoinManager.java

```diff
@@ -233,25 +235,46 @@ private void executeJoinRequest(JoinRequest joinRequest, Connection connection)
         }
     }
 
-    private boolean checkClusterStateBeforeJoin(Address target) {
+    private boolean checkClusterStateBeforeJoin(Address address, String uuid) {
         ClusterState state = clusterStateManager.getState();
         if (state == ClusterState.IN_TRANSITION) {
-            logger.warning("Cluster state is in transition process. Join is not allowed for now -> "
+            logger.warning("Cluster state is in transition process. Join is not allowed until "
+                    + "transaction is completed -> "
                     + clusterStateManager.stateToString());
             return true;
         }
 
-        if (state != ClusterState.ACTIVE) {
-            if (!clusterService.isMemberRemovedWhileClusterIsNotActive(target)) {
-                String message = "Cluster state either is locked or doesn't allow new members to join -> "
-                        + clusterStateManager.stateToString();
-                logger.warning(message);
-                OperationService operationService = nodeEngine.getOperationService();
-                operationService.send(new BeforeJoinCheckFailureOperation(message), target);
+        if (state == ClusterState.ACTIVE) {
+            return false;
+        }
+
+        if (clusterService.isMemberRemovedWhileClusterIsNotActive(address)) {
+            MemberImpl memberRemovedWhileClusterIsNotActive =
+                    clusterService.getMemberRemovedWhileClusterIsNotActive(uuid);
+
+            if (memberRemovedWhileClusterIsNotActive != null
+                    && !address.equals(memberRemovedWhileClusterIsNotActive.getAddress())) {
+
+                logger.warning("Uuid " + uuid + " was being used by " + memberRemovedWhileClusterIsNotActive
+                        + " before. " + address + " is not allowed to join with a uuid which belongs to"
+                        + " a known passive member.");
+
                 return true;
             }
+
+            return false;
         }
-        return false;
+
+        if (clusterService.isMemberRemovedWhileClusterIsNotActive(uuid)) {
```

Comment:
- When cluster is active, we allow anyone to join.
- When cluster is not active, we only allow only known members (members those are removed while not active) to join. A member will be known if;
    * either its address is known, then we expect that;
        - either its uuid is unknown 
        - or address mapped to that uuid is equal to its address, means its returning using the same address and uuid
    * or only its uuid is known

For example, assume `[A, Uuid-1]`, `[B, Uuid-2]` are known when cluster is not active. Then we allow;
- `[A, Uuid-1]` 
- `[B, Uuid-3]`
- `[C, Uuid-1]`

to join. But we don't allow;

- `[A, Uuid-2]`
- `[B, Uuid-1]` 
to join.